### PR TITLE
Iss223 -- deaths from injury

### PR DESCRIPTION
**Pull request by vpancini**

### CODE REVIEW

**Comment by vpancini** [comment]
Data used for this metric are available on [Box](https://urbanorg.box.com/s/tulmxqirl015c4l1eivs2v68dwrjx7rx)

File: 04_health/exposure-to-trauma/final/exposure-to-trauma-2020.csv
Comment (wcurrangroome) [inline 04_health/exposure-to-trauma/final/exposure-to-trauma-2020.csv:1]: It looks like this is the output file that was generated from a prior run of the script when the file output name was different. This shouldn't be included in the commit that's merged into the base branch.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd
Comment (wcurrangroome) [inline 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:1]: Any reason for .Rmd over .qmd? If possible (it should be pretty easy), probably worthwhile to use .qmd over .Rmd at this point.

**Review by wcurrangroome — CHANGES_REQUESTED** [review]
Minor suggestions throughout, on the whole looks great! I think the primary outstanding item is to figure out and implement an approach to crosswalking a couple of geographies that don't align between CDC wonder data and our universe of counties. Can you check with Aaron and JP (and/or whomever the right folks are) about guidance on this? If nothing existing, the `interpolate_pw()` function from library(tidycensus) is likely a fine approach.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 1]
```diff
@@ -0,0 +1,434 @@
...
+  + States
+  + States: *All* (The United States)
+  + 2013 Urbanization: All Categories
+3. Select demographics:
+  + Ten-Year Age Groups: All Ages
+  + Gender: All Genders
+  + Hispanic Origin: All Origins
+  + Single Race 6: All Races
+4. Select year and month:
+  + Year/Month: 2016. 2017, 2018, 2019, 2020
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:109]: Selecting `2016, 2017, 2018, 2019, 2020` creates a 5-year pool that is used as the data for 2020 only. (To get the data for 2019, the user would select the years `2015, 2016, 2017, 2018, 2019`. I will be more explicit in my instruction

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 2]
```diff
@@ -0,0 +1,434 @@
...
+  + Precision: 1 decimal places
+  + Data Access Timeout: 10 minutes
+  
+The specifications for the 2020 extract are saved [here](https://wonder.cdc.gov/controller/saved/D76/D377F651). Note that counties with fewer than ten deaths (for all the ICD-10 codes specified) are suppressed. Population figures are also suppressed when the population represents fewer than ten persons. Mortality rates are marked as "unreliable" when the death count is less than 20. Rates are marked as "not applicable" when the population denominator figure is unavailable, such as persons of "not stated" or unknown age or Hispanic origin.  
+
+Once the extract is created, select "Export" to generate a .txt file. Move the .txt file into the appropriate directory to read in the data.
+
+```{r}
+# List all the files in the `bridged-race` folder
+paths <- list.files(here::here("04_health", "exposure-to-trauma", "data", "bridged-race"),
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:143]: Data stored on Box will not be accessible to the public. Users will have to manually download the data. I will make the instructions for that section more clear.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 3]
```diff
@@ -0,0 +1,434 @@
...
+
+```{r setup}
+options(scipen = 999)
+
+library(urbnthemes)
+library(tidyverse)
+library(readxl)
+library(writexl)
+
+library(rvest)
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:34]: Good call. I always intend to do this and usually forget. Most of those packages were used in last year's update of the metric to read in the data from a different source.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 4]
```diff
@@ -0,0 +1,434 @@
...
+```{r}
+# Download file directly from internet
+download.file(
+  url = "https://www.countyhealthrankings.org/sites/default/files/media/document/2022%20County%20Health%20Rankings%20Data%20-%20v1.xlsx",
+  destfile = here::here("04_health/exposure-to-trauma/data/chr2022.xlsx"),
+  mode = "wb"
+)
+
+# Read in CHR data
+chr2022 <- readxl::read_excel(here::here("04_health", "exposure-to-trauma", "data", "chr2022.xlsx"),
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:270]: Setting `guess_max = Inf` produced a different warning: `Warning: `guess_max` is a very large value, setting to `21474836` to avoid exhausting memoryNew names:` - but I still think this is better than the previous warning which was `Warning: Expecting logical in Q3172 / R3172C17: got 'x'New names:`.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 5]
```diff
@@ -0,0 +1,434 @@
...
+  scatter_grid()
+
+```
+
+
+## 5. Save the data
+This section reads in the Urban Institute's universe of counties, joins the CDC injury death data to those counties, and writes out the file.
+
+Read in Urban county file for 2014-2020
+```{r}
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:341]: Done. Thanks for fixing that!

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 6]
```diff
@@ -0,0 +1,434 @@
...
+
+This table shows the injury death rates in each county. Click on the variable columns to sort the table. 
+    todo(): could not get `reactable()` to work, but it worked for previous version of this metric
+```{r table, echo = FALSE}
+# deaths %>%
+#   select(
+#     county_name, GEOID, year,
+#     rate_injury_death
+#   ) %>%
+#   reactable(
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:208]: I needed to update a dependency for `library(reactable)` and then it worked.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 7]
```diff
@@ -0,0 +1,434 @@
...
+
+library(urbnthemes)
+library(tidyverse)
+library(readxl)
+library(writexl)
+
+library(rvest)
+library(glue)
+library(fs)
+# library(reactable)
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:37]: I needed to update a dependency for `library(reactable)` and then it worked.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 8]
```diff
@@ -0,0 +1,434 @@
...
+```{r}
+missing_urban <- anti_join(x = deaths, y = all_counties, by = "GEOID")
+
+missing_urban %>%
+  group_by(county_name) %>%
+  count()
+
+```
+
+Data for Bedford city, VA (FIPS code 51515) are only available for 2013 and earlier; for computational purposes, counts and population estimates for 2014 and later are set to zero.
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:370]: I moved the section that reads in Urban's universe of counties and joins the CDC WONDER onto it to the beginning. The missing death rates in this section aren't available because these counties no longer exist as such and have been included in the data only for consistency. I included more context explaining this and have removed them from the data.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 9]
```diff
@@ -0,0 +1,434 @@
...
+
+```
+
+```{r}
+rm(test, chr_test, chr2022)
+```
+
+
+### Visualizations of CDC WONDER data
+```{r}
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:316]: I removed observations with missing population counts (the Urban universe of counties is missing population data for 2014) and changed the scale of the x axis to thousands.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 10]
```diff
@@ -0,0 +1,434 @@
...
+                        select(GEOID, deaths, population, rate_injury_death_crude, rate_injury_death),
+                      by = "GEOID") %>%
+  mutate(rate_injury_death_chr = round(rate_injury_death_chr, 1))
+
+```
+
+### Compare values from CHR and CDC WONDER
+
+Visually inspecting the data shows that, for most observations, the rate produced by CHR matches the crude rate from CDC WONDER. There are 161 observations that do not match. These are observations where the rate was `unreliable` because the death count was less than 20. CHR produced the crude rate by dividing the death count by population for these counties and multiplying by 100,000. This computation does not work for us because we are reporting age-adjusted rates, not crude rates.
+```{r}
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:303]: I printed the first 10 rows of data to show that these CHR observations have a death count value of less than 20. I then used a `stopifnot()` function to illustrate that this is true for all these observations. I was more explicit in my explanation that the CDC WONDER documentation states that they mark the rate as `unreliable` when the death count is less than 20, while CHR chose to calculate the rate for those observations anyway.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 11]
```diff
@@ -0,0 +1,434 @@
...
+    rate_injury_death_se = Age.Adjusted.Rate.Standard.Error
+  )
+
+# Remove empty rows resulting from deleting the `Notes` column, and remove the total row
+deaths_14_20 <- deaths_14_20 %>%
+  filter(!is.na(GEOID))
+
+# Change all variables to numeric and replace suppressed or unreliable with missing
+deaths_14_20 <- deaths_14_20 %>%
+  # mutate_all(~ifelse(. %in% c("Suppressed", "Unreliable"), NA, .)) %>%
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:178]: I agree the commented out approach is better. I initially went with the currently un-commented out approach because it replaced those text values with `NA` AND converted the rates and counts from character variables to numeric. With the commented out approach, I had to convert the variables in a separate step, so I initially chose not to use that approach. However, I agree it is clearer and have gone with the commented out approach instead.

File: 04_health/exposure-to-trauma/final/exposure-to-trauma-2020.csv
Comment (vpancini) [reply 04_health/exposure-to-trauma/final/exposure-to-trauma-2020.csv:1]: Oops. Thanks for flagging. Fixing for next PR.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:1]: No. Have switched to .qmd.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 12]
```diff
@@ -0,0 +1,434 @@
...
+    - Y87 (Sequelae of intentional self-harm, assault and events of undetermined intent)
+    - Y89 (Sequelae of other external causes)
+7. Other options:
+  + Show Totals
+  + Show Zero Values
+  + Show Suppressed Values
+  + Precision: 1 decimal places
+  + Data Access Timeout: 10 minutes
+  
+The specifications for the 2020 extract are saved [here](https://wonder.cdc.gov/controller/saved/D76/D377F651). Note that counties with fewer than ten deaths (for all the ICD-10 codes specified) are suppressed. Population figures are also suppressed when the population represents fewer than ten persons. Mortality rates are marked as "unreliable" when the death count is less than 20. Rates are marked as "not applicable" when the population denominator figure is unavailable, such as persons of "not stated" or unknown age or Hispanic origin.
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:137]: I've clarified these instructions

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 13]
```diff
@@ -0,0 +1,434 @@
...
+  
+The specifications for the 2020 extract are saved [here](https://wonder.cdc.gov/controller/saved/D76/D377F651). Note that counties with fewer than ten deaths (for all the ICD-10 codes specified) are suppressed. Population figures are also suppressed when the population represents fewer than ten persons. Mortality rates are marked as "unreliable" when the death count is less than 20. Rates are marked as "not applicable" when the population denominator figure is unavailable, such as persons of "not stated" or unknown age or Hispanic origin.  
+
+Once the extract is created, select "Export" to generate a .txt file. Move the .txt file into the appropriate directory to read in the data.
+
+```{r}
+# List all the files in the `bridged-race` folder
+paths <- list.files(here::here("04_health", "exposure-to-trauma", "data", "bridged-race"),
+                    pattern = "[.]txt$", full.names = TRUE)
+
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:145]: Included this check

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 14]
```diff
@@ -0,0 +1,434 @@
...
+deaths <- deaths %>%
+  mutate(
+    county_name = if_else(county_name == "Wade Hampton Census Area, AK", "Kusilvak Census Area, AK", county_name),
+    GEOID = if_else(GEOID == "02270", "02158", GEOID)
+  )
+
+```
+
+Chugach Census Area, AK and Copper River Census Area, AK are still missing mortality data from CDC. In early 2019, Valdez-Cordova Census Area, AK was abolished and replaced by the Chugach Census Area and the Copper River Census Area. 
+    todo(): not sure if we already have a way to crosswalk this
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:397]: I implemented this for the age-adjusted rate. My understanding is that this approach won't work for interpolating the confidence interval or standard error because the population is smaller in each of Chugach Census Area and Copper River Census Area than in Valdez-Cordova Census Area.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 15]
```diff
@@ -0,0 +1,434 @@
...
+final_data <- left_join(x = all_counties,
+                          y = deaths %>%
+                            select(
+                              year, GEOID, rate_injury_death, rate_injury_death_lb, rate_injury_death_ub, rate_injury_death_se, rate_injury_death_quality
+                            ),
+                          by = c("GEOID", "year")
+                          )
+
+# Check for observations that did not match
+test <- anti_join(all_counties, deaths, by = c("GEOID", "year"))
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:420]: Included where necessary. Also loaded the `tidylog` package at the beginning which helps with some of this.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 16]
```diff
@@ -0,0 +1,434 @@
...
+
+# Remove obsolete variables
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:422]: No, and this whole step was rearranged.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 17]
```diff
@@ -0,0 +1,434 @@
...
+final_data <- final_data %>%
+  select(-GEOID)
+
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:425]: Added where necessary

File: .gitignore [hunk 18]
```diff
@@ -1,3 +1,14 @@
+vincent
```
[the remark this answers is not in the document]
Comment (vpancini) [reply .gitignore:1]: It doesn't, I was just following existing convention that I noticed in the file (other people seemed to write their name above the files they added to the `.gitgnore` document. Have deleted.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd [hunk 19]
```diff
@@ -0,0 +1,434 @@
...
+    # rate_injury_death_lb = Crude.Rate.Lower.95..Confidence.Interval,
+    # rate_injury_death_ub = Crude.Rate.Upper.95..Confidence.Interval,
+    # rate_injury_death_se = Crude.Rate.Standard.Error,
+    rate_injury_death = Age.Adjusted.Rate,
+    rate_injury_death_lb = Age.Adjusted.Rate.Lower.95..Confidence.Interval,
+    rate_injury_death_ub = Age.Adjusted.Rate.Upper.95..Confidence.Interval,
+    rate_injury_death_se = Age.Adjusted.Rate.Standard.Error
+  )
+
+# Remove empty rows resulting from deleting the `Notes` column, and remove the total row
```
[the remark this answers is not in the document]
Comment (vpancini) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd:172]: Yes it does. I have clarified.

File: .gitignore [hunk 20]
```diff
@@ -1,3 +1,14 @@
...
+04_health/county-populations.csv
+04_health/exposure-to-trauma/data/
+04_health/access-health-services*
+07_safety*
+geographic-crosswalks*
+04_health/exposure-to-trauma/cdc2020.txt
+04_health/exposure-to-trauma/cdc2021_single_race.txt
+04_health/exposure-to-trauma/chr2023.xlsx
+
  kevin
```
[the remark this answers is not in the document]
Comment (vpancini) [reply .gitignore:9]: `04_health/exposure-to-trauma/chr2023.xlsx` is still relevant but have removed the other two lines.

**Comment by vpancini** [comment]
I have addressed all reviewer comments and pushed the updated files.

File: 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd (hunk 14 above)
[the remark this answers is not in the document]
Comment (wcurrangroome) [reply 04_health/exposure-to-trauma/exposure-to-trauma-2020.Rmd]: Thanks for trying this! I should have thought this through more carefully before making this suggestion; I don't think this is working correctly. I think the fundamental challenge is that the metric we're reporting (death rates) is standardized by population. So when we interpolate population (count) and deaths (count) and then calculate the deaths rates for each of the new census areas... we're just back at the original death rate we started with. My suggestion would be to just set the death rate for these new census areas to be the same and to set the data quality flag for these observations as low since I'm not sure what else we can do here. But I'll link us up with Aaron in a thread to double check.

**Comment by vpancini** [comment]
> * There's one chunk with a testthat assertion that isn't fully built out (maybe it still works as-is?). If you search for "Check that these observations are equal" you should find it; the full code block might look like: `testthat::expect_equal( not_missing$rate_injury_death_chr, not_missing$rate_injury_death_crude)`
> * At "4.3 Identify counties in CDC Wonder but not in Urban universe": I'm seeing 36 observations, but the text refers to 50. Can you double-check this? I'm running things interactively, so I might be getting a different number because I've changed a variable relative to the value it would have at this point in the execution of the script (though I don't think that's the issue).
> * Per discussion, address the two AK census areas.
> * Then, run one last time top-to-bottom and this should be good to go!

- Did you maybe misread this line? This is what I'm seeing in my code:
`# Check that these observations are equal to the CHR observations
testthat::expect_equal(
  not_missing$rate_injury_death_chr,
  not_missing$rate_injury_death_crude
)`
- Good catch. There were 50 observations initially, but after addressing the counties in Urban universe but not in CDC WONDER in section 4.2, this drops to 36. I have updated the text accordingly.
- Addressed the two AK census areas. They now have the same rate and a quality value of `3` to indicate low quality
- Ran everything one chunk at a time to check and then rendered the file!

**Review by wcurrangroome — APPROVED** [review]
